security-oauth2/src/main/java/io/micronaut/security/oauth2/endpoint/token/request/password/ReactiveOpenIdPasswordAuthenticationProvider.java

```diff
@@ -0,0 +1,114 @@
+
+    private final TokenEndpointClient tokenEndpointClient;
+    private final SecureEndpoint secureEndpoint;
+    private final OauthClientConfiguration clientConfiguration;
+    private final OpenIdProviderMetadata openIdProviderMetadata;
+    private final OpenIdAuthenticationMapper openIdAuthenticationMapper;
+    private final ReactiveOpenIdTokenResponseValidator<JWT> tokenResponseValidator;
+
+    /**
+     * @param clientConfiguration        The client configuration
+     * @param openIdProviderMetadata     The provider metadata
+     * @param tokenEndpointClient        The token endpoint client
+     * @param openIdAuthenticationMapper The user details mapper
+     * @param tokenResponseValidator     The token response validator
+     */
+    public ReactiveOpenIdPasswordAuthenticationProvider(OauthClientConfiguration clientConfiguration,
+                                                        OpenIdProviderMetadata openIdProviderMetadata,
+                                                        TokenEndpointClient tokenEndpointClient,
+                                                        OpenIdAuthenticationMapper openIdAuthenticationMapper,
+                                                        ReactiveOpenIdTokenResponseValidator<JWT> tokenResponseValidator) {
+        this.tokenEndpointClient = tokenEndpointClient;
+        this.clientConfiguration = clientConfiguration;
+        this.openIdProviderMetadata = openIdProviderMetadata;
+        this.openIdAuthenticationMapper = openIdAuthenticationMapper;
+        this.tokenResponseValidator = tokenResponseValidator;
+        this.secureEndpoint = getTokenEndpoint(openIdProviderMetadata);
+    }
+
+    @Override
+    public Publisher<AuthenticationResponse> authenticate(T requestContext, AuthenticationRequest<I, S> authenticationRequest) {
+        OpenIdPasswordTokenRequestContext openIdPasswordTokenRequestContext = new OpenIdPasswordTokenRequestContext(authenticationRequest, secureEndpoint, clientConfiguration);
+        return Mono.from(tokenEndpointClient.sendRequest(openIdPasswordTokenRequestContext))
+                .flatMap(openIdTokenResponse ->
+                        Mono.from(tokenResponseValidator.validate(clientConfiguration, openIdProviderMetadata, openIdTokenResponse, null))
+                                .flatMap(jwt -> {
+                                    try {
+                                        OpenIdClaims claims = new JWTOpenIdClaims(jwt.getJWTClaimsSet());
+                                        return Mono.from(openIdAuthenticationMapper.createAuthenticationResponse(clientConfiguration.getName(), openIdTokenResponse, claims, null));
+                                    } catch (ParseException e) {
+                                        // Should never happen as validation succeeded
+                                        return Mono.error(e);
+                                    }
+                                })
+                );
```

Comment:
Using `Flux` instead of `Mono` here might slightly simplify the call stack, as we would be able to remove the last wrapper call completely and just return the Publisher.

Reply:
@jeremyg484 please, create a PR with this one as based with your recommended changes